python/pyiceberg/catalog/jdbc.py

```diff
@@ -0,0 +1,475 @@
+INSERT_NAMESPACE_PROPERTIES_SQL = f"INSERT INTO {NAMESPACE_PROPERTIES_TABLE_NAME} ({CATALOG_NAME}, {NAMESPACE_NAME}, {NAMESPACE_PROPERTY_KEY}, {NAMESPACE_PROPERTY_VALUE}) VALUES "
+INSERT_PROPERTIES_VALUES_BASE = f"(%s,%s,%s,%s)"
+LIST_ALL_PROPERTY_NAMESPACES_SQL = f"SELECT DISTINCT {NAMESPACE_NAME} FROM {NAMESPACE_PROPERTIES_TABLE_NAME} WHERE {CATALOG_NAME} = %s"
+DELETE_NAMESPACE_PROPERTIES_SQL = f"DELETE FROM {NAMESPACE_PROPERTIES_TABLE_NAME} WHERE {CATALOG_NAME} = %s AND {NAMESPACE_NAME} = %s AND {NAMESPACE_PROPERTY_KEY} IN "
+DELETE_ALL_NAMESPACE_PROPERTIES_SQL = f"DELETE FROM {NAMESPACE_PROPERTIES_TABLE_NAME} WHERE {CATALOG_NAME} = %s AND {NAMESPACE_NAME} = %s"
+UPDATE_NAMESPACE_PROPERTIES_START_SQL = f"UPDATE {NAMESPACE_PROPERTIES_TABLE_NAME} SET {NAMESPACE_PROPERTY_VALUE} = CASE"
+UPDATE_NAMESPACE_PROPERTIES_END_SQL = f" END WHERE {CATALOG_NAME} = %s AND {NAMESPACE_NAME} = %s AND {NAMESPACE_PROPERTY_KEY} IN "
+
+
+GET_NAMESPACE_PROPERTIES_SQL = f"SELECT {NAMESPACE_NAME} FROM {NAMESPACE_PROPERTIES_TABLE_NAME} WHERE {CATALOG_NAME} = %s AND {NAMESPACE_NAME} LIKE %s LIMIT 1"
+GET_ALL_NAMESPACE_PROPERTIES_SQL = f"SELECT * FROM {NAMESPACE_PROPERTIES_TABLE_NAME} WHERE {CATALOG_NAME} = %s AND {NAMESPACE_NAME} = %s"
+
+# Custom SQL not from JDBCCatalog.java
+LIST_ALL_NAMESPACES_SQL = f"""
+SELECT DISTINCT ns FROM
+(
+    SELECT {TABLE_NAMESPACE} AS ns FROM {CATALOG_TABLE_NAME} 
+    WHERE {CATALOG_NAME} = %s 
+    UNION
+    SELECT {NAMESPACE_NAME} AS ns FROM {NAMESPACE_PROPERTIES_TABLE_NAME} 
+    WHERE {CATALOG_NAME} = %s
+) AS all_catalog_namespaces
+"""
+
+def _sqlite(**properties: str) -> Any:
+    parsed_uri = urlparse(properties.get("uri"))
+    return sqlite3.connect(database=parsed_uri.path, uri=False)
+
+def _postgresql(**properties: str) -> Any:
+    parsed_uri = urlparse(properties.get("uri"))
+    postgresql_props = {
+        "user": parsed_uri.username,
+        "password": parsed_uri.password,
+        "dbname": parsed_uri.path[1:],
+        "host": parsed_uri.hostname,
+        "port": parsed_uri.port,
+    }
+
+    return db.connect(**postgresql_props)
+
+SCHEME_TO_DB = {
+    "file": _sqlite,
+    "postgresql": _postgresql,
+}
+
+class JDBCCatalog(Catalog):
+    def __init__(self, name: str, **properties: str):
+        super().__init__(name, **properties)
+
+        # Get a database connection for a specific scheme.
+        uri = urlparse(self.properties.get("uri"))
+        uri_scheme = str(uri.scheme)
+        if uri_scheme not in SCHEME_TO_DB:
+            raise ValueError(f"No registered database for scheme: {uri_scheme}")
+        self._get_db_connection = SCHEME_TO_DB[uri_scheme]
+
+    def initialize_catalog_tables(self) -> None:
+        with self._get_db_connection(**self.properties) as conn:
+            try:
+                curs = conn.cursor()
```

Comment:
I'm a big fan of the context manager. How do you feel about using `with conn.cursor() as curs:`

Reply:
I am totally with you.  That's what I had originally and then I tried to make the code portable between sqlite and postgres.  I ran into issues with using context manager with cursors in sqlite, so I decided to close it manually.  Turns out that having something portable is more of a PITA than I thought since they don't share a common way to express parameters in a statement (Postgres uses `%s` and sqlite `?`).  I'm not sure if we want to go down that route, but if not, I am all for using a context manager.  I'll proceed with that change and if we ever want to support sqlite, we can revisit this.